service/service.go

```diff
@@ -52,17 +73,64 @@ func (svc *Service) Start() {
 
 				message.Commit()
 				log.Debug("message committed", log.Data{"instance_id": instanceID})
-			case errorMessage := <-svc.Consumer.Errors():
-				log.Error(fmt.Errorf("aborting"), log.Data{"message_received": errorMessage})
-				svc.Consumer.Closer() <- true
-				svc.Producer.Closer() <- true
-				exitCh <- true
-				return
-			case <-healthCh:
+			case consumerError := <-svc.Consumer.Errors():
+				log.Error(fmt.Errorf("aborting consumer"), log.Data{"message_received": consumerError})
+				close(done)
+			case producerError := <-svc.Producer.Errors():
+				log.Error(fmt.Errorf("aborting producer"), log.Data{"message_received": producerError})
+				close(done)
+			case <-apiErrors:
+				log.Error(fmt.Errorf("server error forcing shutdown"), nil)
+				close(done)
 			}
 		}
 	}()
-	<-exitCh
 
-	log.Info("service dimension extractor stopped", nil)
+	select {
+	case <-done:
+		log.Debug("Quitting after done was closed", nil)
+	case <-signals:
+		log.Debug("Quitting after os signal received", nil)
+	}
+
+	childContext, childCancel := context.WithTimeout(ctx, svc.Shutdown)
+
+	waitGroup := int32(0)
+
+	atomic.AddInt32(&waitGroup, 1)
+	go func() {
+		log.Debug("Attempting to stop listening to kafka consumer group", nil)
+		svc.Consumer.StopListeningToConsumer(childContext)
+		log.Debug("Successfully stopped listening to kafka consumer group", nil)
+		listeningToConsumerGroupStopped <- true
+		<-readyToCloseOutboundConnections
+		log.Debug("Attempting to close http server", nil)
+		api.Close(childContext)
+		log.Debug("Successfully closed http server", nil)
+		log.Debug("Attempting to close kafka producer", nil)
+		svc.Producer.Close(childContext)
+		log.Debug("Successfully closed kafka producer", nil)
+		log.Debug("Attempting to close kafka consumer group", nil)
+		svc.Consumer.Close(childContext)
+		log.Debug("Successfully closed kafka consumer group", nil)
+		atomic.AddInt32(&waitGroup, -1)
+
+		log.Debug("Closed kafka successfully", nil)
+	}()
+
+	// setup a timer to zero waitGroup after timeout
+	go func() {
+		<-time.After(svc.Shutdown)
```

Comment:
this can be replaced by `<-childContext.Done()` since that's already timing the shutdown